syntax = "proto3";

service Calculator {
    //Calculates the result of a equation
    rpc Calculate (CalculationRequest) returns (CalculationResponse) {
    }

}

service CalculatorSlave {
    //Message to execute a calculation on a server slave node
    rpc DoOperation (OperationRequest) returns (OperationResponse) {
    }
}

// The request message to the server master
message CalculationRequest {
    string expression = 1;
}

// The response message of the server master
message CalculationResponse {
    float result = 1;
    enum Status {
        SUCCESS = 0;
        ZERO_DIVISION_ERROR = 1;
        INVALID_FORMAT_ERROR = 2;
    }
    Status status = 2;
}

// The request message to the server slave
message OperationRequest {
    float a = 1;
    string op = 2;
    float b = 3;
}

// The response message of the server slave
message OperationResponse {
    float result = 1;
    enum Status {
        SUCCESS = 0;
        ZERO_DIVISION_ERROR = 1;
        INVALID_FORMAT_ERROR = 2;
    }
    Status status = 2;
}

